src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -302,7 +364,14 @@ def main():
         raise Exception("Podman service failed to start") from e
     else:
         # set up conf and start callback server
-        conf = Conf(job_dir, output_dir, workers, kb_base_url, token_filepath)
+        conf = Conf(
+            job_dir,
+            output_dir,
+            collection_source_dir,
+            workers,
+            kb_base_url,
+            token_filepath,
+        )
```

Comment:
I forgot one thing - currently in the object loop below any files that have already been downloaded will be skipped. However, the downloader still needs to ensure they're linked from the collection

Reply:
Based on https://github.com/kbase/collections/pull/295#discussion_r1187870747, it seems to me part 2 of downloader task is only optional. If that's the case, ensuring they're linked from the collection shouldn't be a mandatory check.

Reply:
Add check_linking_status_for_existing_object logic!

Reply:
I would restructure your code a bit to make it more understandable / simple. I'd do things in 2 phases:

1) Download all the objects that need to be downloaded. This is a subset of the `list_objects` results
2) Link everything in the list_objects results if they don't already have a link. This happens after the queues have been `join()`ed.

Reply:
For example, currently if the file has been downloaded but isn't already linked, it gets downloaded again, which is not the behavior we want

Reply:
Now do things in 2 phrases!

Reply:
:+1: